instance/utils.py

```diff
@@ -44,6 +44,9 @@
 if TYPE_CHECKING:
     from registration.models import BetaTestApplication  # pylint: disable=cyclic-import, useless-suppression
 
+RETRYABLE_STATUS_CODES = frozenset([400] + list(range(500, 600)))
```

Comment:
- This is limited to OpenStack. Could the variables start with `OPENSTACK_` same as other variables?
- These values are useful for OpenCraft's case, but they aren't good defaults:
  - Retrying on a 400 error seems useless in general (400: bad request. If the request is bad, retrying won't help. Unfortunately in OVH's case, it does)
  - Retrying on a POST can be dangerous, and the original default (retrying only on idempotent methods) is saner
  - So it makes more sense if these are set through https://github.com/open-craft/ansible-secrets/ instead (or ansible-playbooks), so that the defaults are sane and the customizations apply to OpenCraft only. The distinction is mainly to keep the code clean and to separate the different parts. In practical terms it doesn't matter much, because I think we're the only ones using Ocim. Ocim is already heavily specific to OpenCraft but we still do. So @arjunsinghy96 you can also mention other solutions as reviewer
- Wherever you add the variables, could you add a comment nearby explaining the reasoning? (why we retry all methods and even 400 errors), since it's not obvious
- Thinking about whether retrying any other POST call will have unintended consequences…: I guess it won't: if N successive calls to do an operation fail and the N+1 th call works, that's still better than failing (failing was the previous behaviour)

Reply:
@clemente, Thanks for your thoughtful review.

> This is limited to OpenStack. Could the variables start with OPENSTACK_ same as other variables?

Done.

> These values are useful for OpenCraft's case, but they aren't good defaults

I've now moved these variables to Django settings to pick up values from env. And by default, it sets sane values for retry.

> So it makes more sense if these are set through https://github.com/open-craft/ansible-secrets/ instead (or ansible-playbooks)

Do we want to retry on `POST` during production? For now, I've set these environment variables in CircleCI so that integration tests retry on `POST`. But do we want the same behavior on production as well? If no, we don't need to set anything in `ansible-secrets`.

> Thinking about whether retrying any other POST call will have unintended consequences…: I guess it won't: if N successive calls to do an operation fail and the N+1 th call works, that's still better than failing (failing was the previous behaviour)

This is why I am a bit suspicious to set retry on `POST` in production.

Reply:
@shimulch 

> Do we want to retry on POST during production? For now, I've set these environment variables in CircleCI so that integration tests retry on POST. But do we want the same behavior on production as well? If no, we don't need to set anything in ansible-secrets.

Good idea. No change is required in production. I tried the API calls hundreds of times in Ocim stage and production and they weren't flaky, though the same calls were flaky from outside OVH infrastructure. Probably due to network latency. So I think that it's better to do the non-standard 400 retry just in integration tests, and leave production as it is, just in case.

If you change the setting in CircleCI, please write down the new setting in vault (there's a section).

Then we don't need the 400 in this PR, right? And what about the other methods? (PUT, …). The defaults in this PR should be the same as the defaults before this PR (i.e. no behaviour change); only CircleCI will switch on the special 400 (+POST?) retry.

Reply:
@clemente, I've removed 400 from `OPENSTACK_RETRYABLE_STATUS_CODES`. I've set `OPENSTACK_RETRYABLE_STATUS_CODES` and `OPENSTACK_RETRYABLE_METHOD_WHITELIST` in both CircleCI environment and vault (`OpenCraft IM - Integration test settings`).

`PUT` is a default allowed method from [urllib3](https://github.com/urllib3/urllib3/blob/1.26.5/src/urllib3/util/retry.py#L212). We are only adding `POST` to the list during integration testing.